File: src/main/java/org/dukcode/ps/codetree/trail02/chapter06/lesson01/IntroCattleInARowingUp2.java
package org.dukcode.ps.codetree.trail02.chapter06.lesson01;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-cattle-in-a-rowing-up-2/description">LINK</a>
 */
public class IntroCattleInARowingUp2 {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());
    int[] arr = new int[n];
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < n; i++) {
      arr[i] = Integer.parseInt(st.nextToken());
    }

    bw.write(String.valueOf(solve(arr, -1, 0)));

    br.close();
    bw.close();
  }

  private static int solve(int[] arr, int idxBefore, int cnt) {
    if (cnt == 3) {
      return 1;
    }

    int ret = 0;
    for (int idxNow = idxBefore + 1; idxNow < arr.length; idxNow++) {
      if (idxBefore != -1 && arr[idxBefore] > arr[idxNow]) {
        continue;
      }

      ret += solve(arr, idxNow, cnt + 1);
    }

    return ret;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter06/lesson01/IntroCattleInARowingUp2.kt
package org.dukcode.ps.codetree.trail02.chapter06.lesson01.introCattleInARowingUp2

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-cattle-in-a-rowing-up-2/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()
    val heights = parseInts()

    var cnt = 0
    for (i in 0 until n) {
        for (j in i + 1 until n) {
            for (k in j + 1 until n) {
                if (heights[i] <= heights[j] && heights[j] <= heights[k]) {
                    cnt++
                }
            }
        }
    }

    bw.write(cnt.toString())

    bw.flush()
}

private fun solve(
    heights: IntArray,
    lastIdx: Int,
    cnt: Int,
): Int {
    if (cnt == 3) {
        return 1
    }

    return (lastIdx + 1 until heights.size).sumOf {
        if (lastIdx != -1 && heights[lastIdx] > heights[it]) {
            0
        } else {
            solve(heights, it, cnt + 1)
        }
    }
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
